mapper/pchain/tx_builder.go

```diff
@@ -1,6 +1,316 @@
+		return nil, nil, fmt.Errorf("parse memo failed: %w", err)
+	}
+
+	tx := &platformvm.Tx{UnsignedTx: &platformvm.UnsignedAddValidatorTx{
+		BaseTx: platformvm.BaseTx{BaseTx: avax.BaseTx{
+			NetworkID:    sMetadata.NetworkID,
+			BlockchainID: blockchainID,
+			Outs:         outs,
+			Ins:          ins,
+			Memo:         memo,
+		}},
+		Stake: stakeOutputs,
+		Validator: validator.Validator{
+			NodeID: nodeID,
+			Start:  sMetadata.Start,
+			End:    sMetadata.End,
+			Wght:   sumOutputAmounts(stakeOutputs),
+		},
+		RewardsOwner: rewardsOwner,
+		Shares:       sMetadata.Shares,
+	}}
+
+	return tx, signers, nil
+}
+
+func buildAddDelegatorTx(
+	matches []*parser.Match,
+	sMetadata Metadata,
+	codec codec.Manager,
+	avaxAssetID ids.ID,
+) (*platformvm.Tx, []*types.AccountIdentifier, error) {
+	if sMetadata.StakingMetadata == nil {
+		return nil, nil, errInvalidMetadata
+	}
+
+	blockchainID := sMetadata.BlockchainID
+
+	nodeID, err := ids.NodeIDFromString(sMetadata.NodeID)
+	if err != nil {
+		return nil, nil, err
+	}
+	rewardsOwner, err := buildOutputOwner(sMetadata.RewardAddresses, sMetadata.Locktime, sMetadata.Threshold)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ins, _, signers, err := buildInputs(matches[0].Operations, avaxAssetID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse inputs failed: %w", err)
+	}
+
+	outs, stakeOutputs, _, err := buildOutputs(matches[1].Operations, codec, avaxAssetID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse outputs failed: %w", err)
+	}
+
+	memo, err := mapper.DecodeToBytes(sMetadata.Memo)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse memo failed: %w", err)
+	}
+
+	tx := &platformvm.Tx{UnsignedTx: &platformvm.UnsignedAddDelegatorTx{
+		BaseTx: platformvm.BaseTx{BaseTx: avax.BaseTx{
+			NetworkID:    sMetadata.NetworkID,
+			BlockchainID: blockchainID,
+			Outs:         outs,
+			Ins:          ins,
+			Memo:         memo,
+		}},
+		Stake: stakeOutputs,
+		Validator: validator.Validator{
+			NodeID: nodeID,
+			Start:  sMetadata.Start,
+			End:    sMetadata.End,
+			Wght:   sumOutputAmounts(stakeOutputs),
+		},
+		RewardsOwner: rewardsOwner,
+	}}
+
+	return tx, signers, nil
+}
+
+func buildOutputOwner(
+	addrs []string,
+	locktime uint64,
+	threshold uint32,
+) (*secp256k1fx.OutputOwners, error) {
+	rewardAddrs := make([]ids.ShortID, len(addrs))
+	for i, addr := range addrs {
+		addrID, err := address.ParseToID(addr)
+		if err != nil {
+			return nil, err
+		}
+		rewardAddrs[i] = addrID
+	}
+
+	ids.SortShortIDs(rewardAddrs)
+
+	return &secp256k1fx.OutputOwners{
+		Locktime:  locktime,
+		Threshold: threshold,
+		Addrs:     rewardAddrs,
+	}, nil
+}
+
+func buildInputs(
+	operations []*types.Operation,
+	avaxAssetID ids.ID,
+) (
+	ins []*avax.TransferableInput,
+	imported []*avax.TransferableInput,
+	signers []*types.AccountIdentifier,
+	err error,
+) {
+	for _, op := range operations {
+		UTXOID, err := mapper.DecodeUTXOID(op.CoinChange.CoinIdentifier.Identifier)
```

Comment:
I don't think we can as you mentioned this is an offline function. That said I don't think it's necessarily a problem. As long as the signer is allowed to sign this transaction, inputs can be multisig. The balance update will be ignored in the data endpoint, but it will be done so regardless of whether the transaction was constructed with Rosetta or not.

Reply:
Gotcha.